### PR TITLE
Fixed #33350 -- Reallowed using cache decorators with duck-typed HttpRequest.

### PR DESCRIPTION
**Pull request by felixxm**
ticket-33350

Regression in 3fd82a62415e748002435e7bad06b5017507777c.

Thanks Terence Honles for the report.

### CODE REVIEW

File: django/views/decorators/cache.py [hunk 1]
```diff
@@ -48,7 +49,8 @@ def never_cache(view_func):
     """
     @wraps(view_func)
     def _wrapped_view_func(request, *args, **kwargs):
-        if not isinstance(request, HttpRequest):
+        # Don't check HttpRequest instance to allow duck-typing.
+        if not hasattr(request, 'META'):
```
Comment (felixxm) [inline django/views/decorators/cache.py:53]: I really wanted to be smarter here, but couldn't thing of anything better. At worst, we can miss some misuse :shrug:

**Comment by felixxm** [comment]
@terencehonles Does it work for you?

File: tests/decorators/tests.py [hunk 2]
```diff
@@ -493,6 +493,15 @@ def a_view(request):
         self.assertIsNone(r.get('X-Frame-Options', None))
 
 
+class HttpRequestProxy:
+    def __init__(self, request):
+        self._request = request
+
+    def __getattr__(self, attr):
+        """Proxy to the underlying HttpRequest object."""
+        return getattr(self._request, attr)
```
Comment (carltongibson) [inline tests/decorators/tests.py:502]: Good. 👍

File: django/views/decorators/cache.py (hunk 1 above)
Comment (carltongibson) [reply django/views/decorators/cache.py:52]: I'd probably phase the comment in the positive... 

```
            # Ensure argument looks like a request.
            if not hasattr(request, 'META'):
```

I think `META` is a pretty good indicator. (The idea is to catch a moderately common mistake, not _any possible mis-use_)

**Review by carltongibson — APPROVED** [review]
Thanks for picking this up @felixxm! Looks good.

File: django/views/decorators/cache.py (hunk 1 above)
Comment (felixxm) [reply django/views/decorators/cache.py:52]: Updated.

**Comment by felixxm** [comment]
@carltongibson Thanks for checking :+1:

**Review by terencehonles — APPROVED** [review]
Yes, looks good to me. Thanks! I hadn't had a chance to bump the PR I had written.

**Comment by felixxm** [comment]
@terencehonles Thanks :+1:

**Pull request merged**